### PR TITLE
Initilize catalogue API

### PR DESCRIPTION
**Pull request by ErlendHaa**
This is an initial PR for the catalogue API. It's mostly essential
boilerplate and auth needed in order to run the API. Other non-essential
things such as swagger docs is left for later.

This API aims at providing 2 features:

1) Discoverability; The caller can use this API to retrieve metadata
   about Seismic cubes it has access to. This information is stored in a
   database which the API queries on-behalf of the user. It also
   provides the caller with the locations of the cubes (in form of BLOB
   URLs).

2) Issuing SAS tokens; If requested, the API can generate and sign SAS
   tokens that can be used together with the BLOB URL to get access to
   the cubes.

The API is very basic, with a single endpoint to retrieve all cubes the
user have access to. The only customization point is whether the API
should provide the caller with signed URI's (i.e. generate SAS-token)
for each cube in the response.

Architecture
------------

    +--------+           +-----+           +----------+
    | client | ---(1)--> | API | ---(2)--> | database |
    |        | <--(4)--  |     | <--(3)--- |          |
    +--------+           +-----+           +----------+

1) Request with a valid Oauth2 Access Token. The API is configured with
   and Azure App Registration, which will be used for token validation.

2) The API will query a database for cube entries. The API will use
   the value of the Roles claim in the Access Token to determine which
   cubes the user have access to. Each role correspond to a physical
   field.

3-4) Optionally generate user delegation sas token for the returned
  cubes. The "user" is a Service Principle (i.e. the App Reg). Which
  implies that the registered App Reg must have the role
  "Storage Blob Delegator" on storage account you which to be able to
  issue sas tokens from.

### CODE REVIEW

**Review by achaikou — COMMENTED** [review]
Write on my grave "killed by abbrv"
No e2e test run as I never know how to authorize stuff to retrieve tokens. Need help and info.

File: api/internal/auth/auth.go [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+	return &udcProvider{ sp: sp }
+}
+
+func (p *udcProvider) getUserDelegationCredential(
+	storageAccount string,
+	ttl            time.Duration,
+) (*UserDelegationCredential, error) {
+	client, err := service.NewClient(
+		fmt.Sprintf("https://%s.blob.core.windows.net/", storageAccount),
+		p.sp.GetCredential(),
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply api/internal/auth/auth.go:103]: Although I havn't found any explicit rules about it, using a single letter to represent the object itself in methods are common practice in Go. E.g. https://go.dev/doc/effective_go#methods and https://gobyexample.com/methods. Don't hate the player, hate the game.

It makes sense that this is an internal method as it's not mean't to be called directly. 

I agree about `sp` tho, will change to `servicePricipal`

File: api/internal/auth/auth.go [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+	ttl time.Duration,
+) *udcCachingProvider {
+	return &udcCachingProvider{
+		udcProvider: NewUdcProvider(sp),
+		ttl:         ttl,
+		cache:       map[string]UserDelegationCredential{},
+	}
+}
+
+func (u *udcCachingProvider) UdcFunc(account string) (*UserDelegationCredential, error) {
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply api/internal/auth/auth.go:155]: Again, this is how receiver names are written in Go. Please redirect your anger accordingly 😄

File: api/internal/auth/auth.go [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+	u.mutex.Lock()
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply api/internal/auth/auth.go:156]: It guard against a race condition. If requests are reading the cache while it's being updated, the requests will get an corrupted value. Even worse, if multiple requests are writing the same key at the same time we corrupt the value in the cache, and it wont be corrected until the value expires again. Yes, it's rare but we need to guard against race conditions.

It's obviously related to cache, what else would it be related to?

File: api/internal/auth/auth.go [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+
+/** A caching user delegation credential (udc) provider
+ *
+ * Fetching new user delegation credential can be quite time-consuming as it
+ * involves a request to azure. This struct differs from the udcProvider only
+ * in that user delegation credentials are cached between invocations off
+ * udcFunc. Credentials are cached for ttl (time-to-live) time before a fresh
+ * one is fetched from azure.
+ */
+type udcCachingProvider struct {
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply api/internal/auth/auth.go:137]: > Provider: a person or thing that provides something.
It provides you with user delegation credentials. These could be cached or fetched live

File: api/internal/auth/auth.go [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+	sp servicePrinciple
+}
+
+func NewUdcProvider(sp servicePrinciple) *udcProvider {
+	return &udcProvider{ sp: sp }
+}
+
+func (p *udcProvider) getUserDelegationCredential(
+	storageAccount string,
+	ttl            time.Duration,
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply api/internal/auth/auth.go:99]: Completely common term in e.g. networking, and often used for caching: https://en.wikipedia.org/wiki/Time_to_live

File: api/internal/auth/auth.go [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+ * one is fetched from azure.
+ */
+type udcCachingProvider struct {
+	*udcProvider
+	mutex sync.Mutex
+	ttl   time.Duration
+	cache map[string]UserDelegationCredential
+}
+
+func NewUdcCachingProvider(
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply api/internal/auth/auth.go:144]: A very good point, and potentially a big oversight on my part! We need to investigate

File: api/internal/auth/auth.go [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+}
+
+/** Get key/secret from azure key-vault
+ *
+ * This function is pure convinience. It hids some noicy lines needed to fetch
+ * keys from the key vault on server-startup. It will call log.Fatal if for any
+ * reason the request key cannot be retrieved, which means it's not suited
+ * to be called while the server is running.
+ */
+func GetSecretFromKeyVault(sp servicePrinciple, vaultURI, secret string) string {
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply api/internal/auth/auth.go:58]: It's golang syntax?

File: api/api/catalogue.go [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+type cube struct {
+	Country          string `json:"country"          example:"Norway"`
+	Field            string `json:"field"            example:"Volve"`
+	FilenameOnUpload string `json:"filenameOnUpload" example:"myfile.segy"`
+	Url              string `json:"url"              example:"https://account.blob.core.windows.net/cube"`
+}
+
+type Request struct {
+	SasDuration uint `json:"sasDuration" example:"60"`
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply api/api/catalogue.go:25]: Future improvement

File: api/cmd/catalogue/main.go [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+		"Application client ID for azure Service Principle",
+		"string",
+	)
+	getopt.FlagLong(
+		&opts.clientSecret,
+		"client-secret",
+		0,
+		"Application client secret",
+		"string",
+	)
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply api/cmd/catalogue/main.go:57]: I will keep them and document that they could be set with env vars instead

File: api/internal/database/database.go [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+ * implements this interface could be used as a backend for the Catalogue API.
+ */
+type Adapter interface {
+	/** Get Cubes 
+	 *
+	 * Fetches and returns all cubes where CubeEntry.Field is in fields.
+	 * If *includeNonRestricted=True* all cubes with CubeEntry.Restricted=False
+	 * will also be included in the output set, regardless of their Field value
+	 */
+	GetCubes(fields []string, includeNonRestriced bool) ([]CubeEntry, error)
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply api/internal/database/database.go:25]: I remove it, together with field-restricted filtering in the sql query (until we know how it should behave). Meaning we are strict until we now how the country / field restricted plays in

File: api/internal/postgres/postgres.go [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+	sql := fmt.Sprintf(
+		"SELECT * FROM catalogue.cube WHERE field IN (%s)",
+		formatStringArray(fields),
+	)
+
+	if includeNonRestricted {
+		sql += " OR access_field_restricted IS NOT TRUE"
+	}
+	
+	rows, err := c.pool.Query(context.Background(), sql)
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply api/internal/postgres/postgres.go:40]: These are not user-inputed strings, they come from the token. So current implementation is fine

File: api/internal/auth/auth.go [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+
+	udc, err := client.GetUserDelegationCredential(context.Background(), keyInfo, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return &UserDelegationCredential{ cred: *udc, expiresAt: expiry }, nil
+}
+
+func (p *udcProvider) UdcFunc(account string) (*UserDelegationCredential, error) {
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply api/internal/auth/auth.go:125]: Removed

File: api/internal/auth/auth_test.go [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+				"aud"   : "valid audience",
+				"exp"   : time.Now().Add(time.Minute *  2).Unix(),
+				"nbf"   : time.Now().Add(time.Minute *  1).Unix(),
+				"iat"   : time.Now().Add(time.Second * -1).Unix(),
+				"roles" : []string{"Read"},
+			}),
+		},
+		{
+			name:     "Missing 'role' claim",
+			expected: http.StatusOK,
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply api/internal/auth/auth_test.go:140]: Pointless, I remove

File: Dockerfile [hunk 14]
```diff
@@ -0,0 +1,19 @@
+FROM golang:1.18-alpine as builder
+WORKDIR /src/api
+COPY go.mod go.sum ./
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply Dockerfile:3]: Moved docker file to api/

File: api/api/catalogue.go [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+			}
+
+			containerURL = signContainerURL(containerURL, sas)
+		}
+
+		response = append(response, cubeResponse{
+			Country:          cube.Country,
+			Field:            cube.Field,
+			FilenameOnUpload: cube.FilenameOnUpload,
+			Url:              containerURL,
```
Comment (achaikou) [inline api/api/catalogue.go:93]: Btw, we have a very weird design as for our own vds-slice.
Here we return full URL, but our API accepts separately vds and separately sas.

(though with current design users still need to do something more, as this is a link to container, and I currently put vds files into "vds" directory. Unsure if we want to return link to container, or to vds)

Another interesting point:
When I try to use received SAS to read data from vds-slice, I get 403.
But when I use the same service principal in python upload script to generate sas, that sas (with read privileges) can read the data.
Any idea? These SASs work for you?

File: api/cmd/catalogue/main.go [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+
+	connectionString := auth.GetSecretFromKeyVault(
+		servicePrincipal,
+		opts.keyVault,
+	 	"db-reader-connection-string",
+	)
+
+	db, err := postgres.NewAdapter(connectionString)
+	defer db.Close()
```
Comment (achaikou) [inline api/cmd/catalogue/main.go:102]: > should check returned error before deferring db.Close() (SA5001)

File: api/internal/auth/auth.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (achaikou) [reply api/internal/auth/auth.go]: Yep. We need to account for it.
Checked with creating key valid for 1 minute and requesting 1 hour sas.
Stopped working almost immediately as sas has ske which seems to be end-of-key.

File: api/internal/auth/auth.go [hunk 17]
```diff
@@ -0,0 +1,286 @@
...
+
+	resp, err := client.GetSecret(context.TODO(), secret, "", nil)
+	if err != nil {
+		log.Fatal(err)
+	}
+	return *resp.Value
+}
+
+/** user delegation credential (udc) with expiry */
+type UserDelegationCredential struct {
```
[the remark this answers is not in the document]
Comment (achaikou) [reply api/internal/auth/auth.go]: Rant aside, question was: can we rename this into UserDelegationCredentialWithExpiery or to SigningCredential to not confuse with original UserDelegationCredential from service?

**Review by achaikou — COMMENTED** [review]
I still couldn't test it fully with auth.
Don't have access to your SP, and our test one is apparently not configured (not found in tenant) and I am in no mood to find out how to do that today (even though it probably anyway should be a separate one). So next time!

File: api/internal/auth/auth.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply api/internal/auth/auth.go]: No, I don't like any of those names. SigningCredential is less specific. And there is no confusion, they live in different packages so it's always clear with one is which? `auth.UserDelegationCredential` or `service.UserDelegationCredential`. Will keep as is

File: api/internal/auth/serviceprincipal.go [hunk 18]
```diff
@@ -0,0 +1,33 @@
+package auth
+
+import (
+	"log"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+)
+type servicePrincipal struct {
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply api/internal/auth/serviceprincipal.go:9]: Removed

File: api/cmd/catalogue/main.go [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+		servicePrincipal,
+		time.Duration(24 * time.Hour),
+	)
+
+	sasGenerator := auth.NewUserDelegationSasFunc(udcProvider.UdcFunc)
+	catalogue := api.NewCatalogueAPI(db, sasGenerator)
+
+	app := gin.Default()
+	
+	authServer := fmt.Sprintf("https://login.microsoftonline.com/%s/v2.0", opts.tenantId)
```
[the remark this answers is not in the document]
Comment (ErlendHaa) [reply api/cmd/catalogue/main.go:117]: leaving for now then

File: api/api/catalogue.go (hunk 15 above)
Comment (ErlendHaa) [reply api/api/catalogue.go:93]: >Btw, we have a very weird design as for our own vds-slice.
Here we return full URL, but our API accepts separately vds and separately sas.

Good point, should sas be it's own entry then? Or does it even make sense that the url and sas is separated, maybe we should change vds-slice?

> and I currently put vds files into "vds" directory. Unsure if we want to return link to container, or to vds)
Then I add /vds

> When I try to use received SAS to read data from vds-slice, I get 403
They broke the signing again 🤦 Now a different version is the only one working

**Review by achaikou — COMMENTED** [review]
Create issues/comment for unresolved comments?

**Review by achaikou — COMMENTED** [review]
:heart:

**Review by achaikou — APPROVED** [review]

**Pull request merged**